[lua] Twinkbrix

MowFord opened the pull request:
<!-- Remove space and place 'x' mark between square [] brackets or click the checkbox after saving to affirm the following points: -->
<!-- (it should look like this: - [x] I have ...) -->
**_I affirm:_**
- [x] I understand that if I do not agree to the following points by completing the checkboxes my PR will be ignored.
- [x] I understand I should leave resolving conversations to the LandSandBoat team so that reviewers won't miss what was said.
- [x] I have read and understood the [Contributing Guide](https://github.com/LandSandBoat/server/blob/base/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/LandSandBoat/server/blob/base/CODE_OF_CONDUCT.md).
- [x] I have _**tested my code and the things my code has changed**_ since the last commit in the PR and will test after any later commits.

## What does this pull request do?

Cleans up the logic for [Twinkbrix](https://www.bg-wiki.com/ffxi/Twinkbrix). Started with a cherry-pick from ASB then fell down the rabbit hole of getting retail captures and changed basically everything

- The logic of giving key items based on event id is unchanged from ASB (in `onEventFinish`)
  - Note that the cooldown technically needs to be started when the battle is entered, but afaik all other ENMs do it in the npc lua so that is left as a TODO (esp. since the fights aren't implemented yet)
- The trade logic was mostly correct except that trading 2k gil should only teleport you if you have the gate dial.
  - sub logic exists that if you have the operating lever, prompt if you are sure:
  - ![image](https://github.com/LandSandBoat/server/assets/131182600/252db5a0-bacd-464d-a9ab-0d17830a1e53)
- A message is added to the trade logic to revert to the `onTrigger` informational event if you trade the requirements for either key item
- The onTrigger function only ever gives information about the two ENM key items: `SHAFT_GATE_OPERATING_DIAL` and `SHAFT_2716_OPERATING_LEVER`
  - logic has so many possible combinations it felt impossible but after many captures I noticed the pattern
  - Event ID 50 is a generic "you got what you need" message
  - Event ID 53 is a generic "you can't get the operating lever until X time"
  - Event ID 52 has many possible combinations that vary depending on which key items you have/are on cooldown

Final note about trading 2716 gil:
- SE unveiled on a stream that this NPC had a special combination of gil and day to get a higher chance of winning. 
  - It is assumed to be 2716 on lightsday, due to the special message for trading 2716
  - It is highly likely the day is lightsday, due to player reports of winning often. Takes a bit more testing but we've yet to trade 2716 on lightsday and not win (current code in this PR makes the winning number <=13 in the range 2-20 so decent chance of losing still)
  - The 2716 on lightsday still says you need to roll between 2 and 13 so the higher chance is given (in this code) by changing the range of the random number. This matches what we've seen on retail:
    - ![image](https://github.com/LandSandBoat/server/assets/131182600/891b5855-a5bd-44e1-b900-3c172102b268)
    - ![image](https://github.com/LandSandBoat/server/assets/131182600/a8fc849a-ccb9-4465-a6e9-de0575b367eb)


## Steps to test these changes

trade the following things and notice the behavior depending on if you have the right key item/the ENM cooldown is in the future 
- 2k gil
  - if you have the key item: 
![image](https://github.com/LandSandBoat/server/assets/131182600/3181585a-ca60-428f-86ba-cffc6fc89ad0)

- 2716 gil
  - if you don't have key item and aren't on cooldown:  
![image](https://github.com/LandSandBoat/server/assets/131182600/7f5b95d3-5b2b-44d9-b066-2b0d1381920b)

- between 1k and 10k gil
  - same as above, without the special message
- Sylvan Stone item
  - if you don't have key item and aren't on cooldown:
![image](https://github.com/LandSandBoat/server/assets/131182600/a7044d12-1b17-4c17-82a8-dbf42f759fa0)

MowFord committed:
Pull twinkbrix from asb

Co-authored-by: Tiberon <TiberonKalkaz@gmail.com>
Co-authored-by: Abdiah <62350957+dallano@users.noreply.github.com>

WinterSolstice8 commented on scripts/zones/Oldton_Movalpolos/npcs/Twinkbrix.lua:
```diff
@@ -6,48 +6,117 @@
+        -- teleport for mineShaftWarpCost relies on having SHAFT_GATE_OPERATING_DIAL
+        -- but consumes SHAFT_2716_OPERATING_LEVER (after confirming with you)
+        if player:hasKeyItem(xi.ki.SHAFT_2716_OPERATING_LEVER) then
+            player:startEvent(56, 1781, 23, 1757, 177552692, 8, 17407, 15, 0)
```
This looks like an NPC ID, can you confirm it's either you and or/ the goblin? `177552692`

MowFord replied:
npc id is 16822535

![image](https://github.com/LandSandBoat/server/assets/131182600/8a7abc91-d3ee-4003-afd8-9a8987da862c)

Looks like it works fine without the rest of the params. Want to nix them?

Xaver-DaRed replied:
The first parameter after event ID is an item id. Concretely, the item you defined.

MowFord replied:
Sure seems that way.

MowFord replied:
Updated

Xaver-DaRed reviewed (changes requested):
Fist paramter IS the item ID. Use the enum you already defined instead of the raw ID

MowFord commented:
> Fist paramter IS the item ID. Use the enum you already defined instead of the raw ID

Understood, would you like to remove the extra params from event 56 since they don't seem required? I can leave a comment with the info from the capture

MowFord committed:
Twinkbrix clean up from packet captures

- Teleport via 2k gil relies only on having operating dial
- He informs you of ENM cooldowns on trigger before other things
- special considerations trading 2716 gil
- TODO for ENM cooldowns handled at mine entrance (when fights are coded)
